### ISSUE TITLE
CMake configure fails

### ISSUE THREAD

**Issue by blitz-research** [eliemichel/WebGPU-distribution#13]
Hi,

I'm trying to use option A to build a static 'dawn' lib, but it's failing at the very first step after a clean clone:

```
$ cmake -B build-dawn -DWEBGPU_BACKEND=DAWN
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.20348.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.38.33133.0
-- The CXX compiler identification is MSVC 19.38.33133.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.38.33130/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.38.33130/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Using Dawn backend for WebGPU
CMake Error at build-dawn/_deps/webgpu-backend-dawn-src/cmake/FetchDawn.cmake:61 (add_subdirectory):
  The source directory

    C:/Users/marks/dev/WebGPU-distribution/build-dawn/_deps/dawn-src

  does not contain a CMakeLists.txt file.
Call Stack (most recent call first):
  build-dawn/_deps/webgpu-backend-dawn-src/CMakeLists.txt:19 (include)


CMake Error at build-dawn/_deps/webgpu-backend-dawn-src/cmake/FetchDawn.cmake:108 (target_include_directories):
  Cannot specify include directories for target "dawn_utils" which is not
  built by this project.
Call Stack (most recent call first):
  build-dawn/_deps/webgpu-backend-dawn-src/CMakeLists.txt:19 (include)


-- Configuring incomplete, errors occurred!
See also "C:/Users/marks/dev/WebGPU-distribution/build-dawn/CMakeFiles/CMakeOutput.log".
```

I can actually build a static dawn lib using my own horribly hacked CMakeLists.txt, but this looks a lot cleaner and gets rid of the depot_tools dependency so it'd be nice to get it working.

I'm using msvc2022 and git bash on Windows 10 64bit.

Bye!
Mark

**Comment by eliemichel** [comment]
Hello, could you shre the very `WebGPU.cmake` file that you are using? To check the version of the distribution it is trying to fetch. Also try removing `build-dawn/_deps/webgpu-backend-dawn-src` to force redownloading, maybe something went wrong a previous time you ran CMake and there is some leftover of it!

**Comment by blitz-research** [comment]
Hi,

Thanks for th reply! Here is WebGPU.cmake before attempt to configure. After cmake config fails, it disappears! Is this normal or am I building in the wrong dir or something?
```
include(FetchContent)

set(WEBGPU_BACKEND "WGPU" CACHE STRING "Backend implementation of WebGPU. Possible values are WGPU and DAWN (it does not matter when using emcmake)")

if (NOT TARGET webgpu)
	string(TOUPPER ${WEBGPU_BACKEND} WEBGPU_BACKEND_U)

	if (EMSCRIPTEN OR WEBGPU_BACKEND_U STREQUAL "EMSCRIPTEN")

		FetchContent_Declare(
			webgpu-backend-emscripten
			GIT_REPOSITORY https://github.com/eliemichel/WebGPU-distribution
			GIT_TAG        emscripten-v3.1.45
			GIT_SHALLOW    TRUE
		)
		FetchContent_MakeAvailable(webgpu-backend-emscripten)

	elseif (WEBGPU_BACKEND_U STREQUAL "WGPU")

		FetchContent_Declare(
			webgpu-backend-wgpu
			GIT_REPOSITORY https://github.com/eliemichel/WebGPU-distribution
			GIT_TAG        wgpu-5433868
			GIT_SHALLOW    TRUE
		)
		FetchContent_MakeAvailable(webgpu-backend-wgpu)

	elseif (WEBGPU_BACKEND_U STREQUAL "DAWN")

		FetchContent_Declare(
			webgpu-backend-dawn
			GIT_REPOSITORY https://github.com/eliemichel/WebGPU-distribution
			GIT_TAG        dawn-5869
			GIT_SHALLOW    TRUE
		)
		FetchContent_MakeAvailable(webgpu-backend-dawn)

	else()

		message(FATAL_ERROR "Invalid value for WEBGPU_BACKEND: possible values are WGPU and DAWN, but '${WEBGPU_BACKEND_U}' was provided.")

	endif()
endif()
```

The *exact* steps I'm taking (after completely deleting WebGPU-distribution dir) are:
```
~/dev
$ git clone https://github.com/eliemichel/WebGPU-distribution.git
Cloning into 'WebGPU-distribution'...
remote: Enumerating objects: 461, done.
remote: Counting objects: 100% (113/113), done.
remote: Compressing objects: 100% (52/52), done.
remote: Total 461 (delta 66), reused 84 (delta 50), pack-reused 348
Receiving objects: 100% (461/461), 177.38 MiB | 18.83 MiB/s, done.
Resolving deltas: 100% (167/167), done.

~/dev
$ cd WebGPU-distribution

~/dev/WebGPU-distribution (main)
$ cmake -B build-dawn -DWEBGPU_BACKEND=DAWN
....ERROR HERE!...
```

Here is content of WebGPU-distribution dir after cmake fails (looks a bit like dawn repos?):
```
~/dev/WebGPU-distribution (main)
$ ls
AUTHORS             DEPS             README.md            generator/               src/
BUILD.gn            DIR_METADATA     build-dawn/          go.mod                   test/
CMakeLists.txt      Doxyfile         build_overrides/     go.sum                   third_party/
CMakeSettings.json  LICENSE          codereview.settings  go_presubmit_support.py  tint_overrides_with_defaults.gni
CODE_OF_CONDUCT.md  OWNERS           dawn.json            include/                 tools/
CONTRIBUTING.md     PRESUBMIT.py     dawn_wire.json       infra/                   webgpu-cts/
CPPLINT.cfg         README.chromium  docs/                scripts/
```

I'm sure it's something stupid I'm doing (I suck at cmake and 'build systems'!) but I can't figure out what.

Bye!
Mark

**Comment by blitz-research** [comment]
OK, I had another shot at and this and learned a few things but still can't get it to work.

I created a simple CMakeLists.txt that 'fetched' my own fork of the stb lib:
```
cmake_minimum_required(VERSION 3.0.0...3.24 FATAL_ERROR)
project(testproj VERSION 1.0.0)
include(FetchContent)

FetchContent_Declare(
    testcontent
    GIT_REPOSITORY https://github.com/blitz-research/stb.git
    GIT_TAG        master
    GIT_SHALLOW    TRUE
)

FetchContent_MakeAvailable(testcontent)

```
'cmake -B build' and it works, the repos gets cloned to build/_deps!

So I changed the repository URL to WebGPU-distribution and tag to 'dawn-5869'. The first error I had was:
```
Performing download step for 'dawn-populate'
Initialized empty Git repository in C:/Users/marks/dev/test/build/_deps/dawn-src/.git/
  hint: Using 'master' as the name for the initial branch. This default branch name
  hint: is subject to change. To configure the initial branch name to use in all
  hint: of your new repositories, which will suppress this warning, call:
  hint:
  hint:         git config --global init.defaultBranch <name>
  hint:
  hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
  hint: 'development'. The just-created branch can be renamed via this command:
  hint:
  hint:         git branch -m <name>
```
So I did:
```
git config --global init.defaultBranch main
```
Which fixed the problem. And yes, in retrospect this probably wasn't an error....

The next error was:
```
Performing download step for 'dawn-populate'
Initialized empty Git repository in C:/Users/marks/dev/test/build/_deps/dawn-src/.git/
fatal: not a git repository (or any of the parent directories): .git
```
And this is where things get weird. I was just working in a plain 'test' directory, so decided to try git init-ing it.

And things got a bit further, although I ended up in the same situation as above: the contents of 'outer' project were completely overwritten by what appears to be dawn source code and my test CMakeLists.txt file was gone!

I imagine I shouldn't *have* to have the root CMakeLists.txt (or webgpu.cmake) in a git inited repos for this to work? I didn't have to for stb.

And this is where things stand, will keep hacking, but man I hate this stuff...!

Bye,
Mark

**Comment by blitz-research** [comment]
OK, think I know where the problem is: the dawn branch declares it's own DOWNLOAD_COMMAND to clone dawn source code from googlecode.com, whioch appears to be failing somehow - the dawn-src repos is inited by then not found:

```
Initialized empty Git repository in C:/Users/marks/dev/test/build/_deps/dawn-src/.git/
fatal: not a git repository (or any of the parent directories): .git
```

When it then fetches, it goes on to find the parent repos and dumping dawn there instead!

This is probably becasue I'm using gitbash? What are you using on Windows?

Anyway, removing the DOWNLOAD_COMMAND and re-enabling the comented out GIT_REPOSITORY... etc appears to at least fetch dawn-src properly...although it seems to have popped up a 'credentail manager' request at the end, more f#$%ing drama!

I think I'll just with my own crappy build system for now, thanks for taking a look anyway.

**Comment by eliemichel** [comment]
Reopenning because this last point is actually an issue, thx for investigating!

**It should work no matter the terminal!** The use of a custom `DOWNLOAD_COMMAND` is a workaround because CMake's `GIT_SHALLOW ON` option just does not work (it performs a full clone whether the shallow option is on or off).

Which version of CMake do you use? It's weird because it seems that `git init` is called in the right directory, but not `git fetch`... Could you try this alternative command instead?

```
DOWNLOAD_COMMAND
	git -C ${FETCHCONTENT_BASE_DIR}/dawn-src init &&
	git -C ${FETCHCONTENT_BASE_DIR}/dawn-src fetch --depth=1 https://dawn.googlesource.com/dawn chromium/6108 &&
	git -C ${FETCHCONTENT_BASE_DIR}/dawn-src reset --hard FETCH_HEAD
```

**Comment by blitz-research** [comment]
Hi,

OK, I tried again from scratch and it worked first time!

I have no idea what was up last time, I do clearly remember deleting the dir and cloning again several times so I'm completely stumped. I agree it's like gitbash 'forgot' what the current directory was after initing the repos but I can offer no sane explanation for why.

You're right about GIT_SHALLOW not working properly though - I've since adapted your code to my own project only using GIT_SHALLOW, think I'll try your way again. I'll let you know if I run into any problems like the above in the process.

Feel free to close this again, just 'one of those weird things' I guess that seem to be getting more and more common these days!

Bye,
Mark
